src/core/Akka/Event/LogFilter.cs

```diff
@@ -0,0 +1,325 @@
+    /// <param name="content">Usually the fully expanded message content.</param>
+    public abstract LogFilterDecision ShouldKeepMessage(LogFilterType part, string content);
+}
+
+/// <summary>
+/// Uses a regular expression to filter log messages based on their source.
+/// </summary>
+public sealed class RegexLogSourceFilter : LogFilterBase
+{
+    private readonly Regex _sourceRegex;
+
+    public RegexLogSourceFilter(Regex sourceRegex)
+    {
+        _sourceRegex = sourceRegex;
+    }
+
+    public override LogFilterType FilterType => LogFilterType.Source;
+
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Source)
+            return _sourceRegex.IsMatch(content) ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+public sealed class ExactMatchLogSourceFilter : LogFilterBase
+{
+    private readonly string _source;
+    private readonly StringComparison _comparison;
+
+    public ExactMatchLogSourceFilter(string source, StringComparison comparison = StringComparison.OrdinalIgnoreCase)
+    {
+        _source = source;
+        _comparison = comparison;
+    }
+
+    public override LogFilterType FilterType => LogFilterType.Source;
+
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Source)
+            return content == _source ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+public sealed class RegexLogMessageFilter : LogFilterBase
+{
+    private readonly Regex _messageRegex;
+
+    public RegexLogMessageFilter(Regex messageRegex)
+    {
+        _messageRegex = messageRegex;
+    }
+
+    public override LogFilterType FilterType => LogFilterType.Message;
+
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Message)
+            return _messageRegex.IsMatch(content) ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+/// <summary>
+/// Runs inside the logging actor and evaluates if a log message should be kept.
+/// </summary>
+public class LogFilterEvaluator
+{
+    public static readonly LogFilterEvaluator NoFilters = EmptyLogFilterEvaluator.Instance;
+
+    private readonly LogFilterBase[] _filters;
+
+
+    /// <summary>
+    /// "Fast path" indicator - if this is true, we only evaluate log sources and not the message content.
+    /// </summary>
+    public bool EvaluatesLogSourcesOnly { get; }
+
+    public LogFilterEvaluator(LogFilterBase[] filters)
+    {
+        _filters = filters;
+        EvaluatesLogSourcesOnly = filters.All(x => x.FilterType == LogFilterType.Source);
+    }
+
+    public virtual bool ShouldTryKeepMessage(LogEvent evt, out string expandedLogMessage)
```

Comment:
A bit of a perf optimization here - if we're asked to evaluate including / excluding a log based on the content of its message, we have to expand the log earlier than usual. That can actually be a somewhat expensive operation (in the context of millions / second).

Since we have to expand _anyway_ for filtration purposes, we might as well also make this actor responsible for the original log expansion and emit that back out to the `StandardOutLogger` so we only have to do it once. That's why we have the `out` parameter here.